### DISCUSSION TITLE
Provide a more privacy-focused configuration (not capturing detailed devices info)

### DISCUSSION THREAD

**Discussion by dmakarenko** [ory/kratos#3431] [Q&A] [answered]
Dear Ory community, in my team we have been using a self-hosted version of Kratos for over a year and it has been quite a successful endeavour (kudos to the whole Ory team 🙇).

We are now looking at the privacy aspects of it and trying to minimise the amount of personal data we are capturing (in fact we want to capture none of what could be qualified as PII). What we have noticed that there is a list of devices associated with [Sessions](https://pkg.go.dev/github.com/ory/kratos/session#Session) which currently stores an `ip_address` and a `user_agent` properties. A pull request implementing it is #2715 (by @kelkarajay).

I am wondering if Kratos has a configuration setting allowing not to capture those details.

Alternatively if anybody sees a way to anonymise this information I would also appreciate any pointers & suggestions.

Thanks a lot.

Update: asked the same question via Ory Slack [here](https://ory-community.slack.com/archives/C02PSRUGUAV/p1692205982111779)

**Comment by vinckr** [comment]
Hello @dmakarenko 

Can you elaborate a bit more on the privacy issue? 

AFAICT the devices property is only visible to the user themselves and Ory Network (the latter could not serve the user without knowing their IP). The data in `devices` is not stored anywhere longterm and it is part of the payload you send in any kind of request to any kind of website as well. 
Also I think there is no "hard" PII exposed, as IP address is not quite PII. The only sensitive info is probably the location? Which is not exact and based on the IP address. 

I think its an interesting (and not easy) topic, so happy to start a conversation here! 
(Also I am not a lawyer)

Edit: 
Actually I was wrong and the IP is considered PII in the EU (and somewhat in the US):

> The categorization of an IP address as Personally Identifiable Information (PII) can vary depending on the jurisdiction and specific legislation.
> 
> In the European Union, under the General Data Protection Regulation (GDPR), an IP address is considered PII as it is classified as an online identifier for an individual. The GDPR sets strict rules on how companies doing business in the EU or with EU citizens can handle PII, including IP addresses. This means companies are required to take reasonable precautions to protect such data from hackers and allow EU citizens to delete their data upon request [csoonline.com](https://www.csoonline.com/article/571817/what-is-pii-examples-laws-and-standards.html).
> 
> However, in the United States, guidance on whether IP addresses are considered PII is more fragmented and varies by specific statutes in federal circuits. Businesses are advised to be cautious in making conclusions about their collection and use of IP addresses, as it varies by business type, data type, and legal jurisdiction [lexology.com](https://www.lexology.com/library/detail.aspx?g=5648af5d-268c-4561-985c-9fcd2186c0a9).
> 
> Given that IP addresses could potentially be used to identify an individual, particularly when combined with other pieces of data, they should be treated with a similar level of security as other sensitive PII to avoid potential data privacy law violations [upguard.com](https://www.upguard.com/blog/personally-identifiable-information-pii).

**Reply by dmakarenko** [reply]
Hi Vincent,

We are currently relying on a self-hosted deployment of Ory Kratos. And our goal is to avoid storing any user's Personal Data as in some territories/countries we could only store & process PII on servers located physically within the borders of those countries (e.g. the People's Republic of China). And in such cases in order to be compliant but still have an option of hosting our servers in those countries we believe could be achieved if we stop processing any personal data in Kratos.

It might still be a bit of a grey area as it's quite obvious that as long as an IP packet reaches our servers one could say that we are already processing an IP address on our servers.

**Reply by vinckr** [reply]
Hey Dmitry, 
that makes sense though! We also offer [professional services](https://www.ory.sh/professional-services/), maybe that could be an option to book some time with an Ory engineer to go through the details and what a good solution woudl be.

And yea I agree it's a grey area, so I guess legal opinion from an expert would be good too.

**Comment by Pedr0Rocha** [comment]
Hi,
I found this discussion while looking for a way to prevent kratos (self-hosted) from storing `session_devices.ip_address`, `session_devices.user_agent` and `session_devices.location`.

I need to drop those fields for privacy reasons and I understand there's no dedicated configuration for this. Is this something achievable by a pre/post login hook?

Is adding a property to handle this in the configuration file something we can consider?

Thanks!

**Reply by vinckr** [reply, marked as answer]
Hey @Pedr0Rocha 
I think this should not be stored in Kratos by default anyway - Its been a while since I checked this but it should only be visible to the end user and your Kratos deploy while the session is active. 
I don't think you can modify the session contents with an Ory Action webhook 🤔 

In any case if you use Ory Kratos for a commercial use case in production it is recommended to get an Ory Enterprise License (OEL).
With the OEL you get bug and security fixes as they come in and you don't have to wait for the next release, as well as professional support, and the Ory dev team can help you out with special feature requests like this.  
Plus you ensure long-term sustainability of the project and fund ongoing development.
You can [contact Ory here](https://ory.sh/contact) for next steps.    
The OSS (or community) version of Ory Kratos is provided as is without support (beyond basic community support).
